### PR TITLE
Handle nested string columns with no children in contiguous_split.

### PR DESCRIPTION
**Pull request by nvdbaranec**
Fixes a specific corner case:   String columns with no children (a special form of empty string column that can happen) that are nested inside a list (or struct) column.

This would be useful as a 0.17 PR but isn't strictly necessary, since it's pretty late.

Edit:  Updated the fix so that it always includes a record for src/dst buffers, even if they are of size 0 or have null data pointers.  The previous method that only checked the data pointer being null was unclean and didn't handle a particularly strange case that came up with the Spark plugin:  the plugin was reconstructing columns (on the receiver side of a shuffle) that had size 0 but a non-null data pointer.  This is technically legal but super weird.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by harrism** [comment]
If this PR were delayed until 0.18, is the bug it leaves bad enough that we would need to hotfix release 0.17?

I'm assuming no since there is no issue filed. In that case, please target branch-0.18 with this PR, it's way too late.

**Review by karthikeyann — APPROVED** [review]

**Comment by nvdbaranec** [comment]
> 
> 
> If this PR were delayed until 0.18, is the bug it leaves bad enough that we would need to hotfix release 0.17?
> 
> I'm assuming no since there is no issue filed. In that case, please target branch-0.18 with this PR, it's way too late.

It's a somewhat niche case (empty string columns with no children, nested inside a list or struct column) that would cause an exception to get thrown.   There was a bug merged yesterday that addressed a similar issue (https://github.com/rapidsai/cudf/pull/6853) but I realized after the fact that it doesn't catch this slightly more unusual case.

**Comment by nvdbaranec** [comment]
Adding the do not merge tag for now until Mark sees updated comments.

**Comment by nvdbaranec** [comment]
Roping in @sperlingxx  to sanity check some changes that apply to his original bug fix.

Fundamentally : I changed it so that the logic in the buf_info functor matches the logic of the related functions (like setup_src_buf_data and count_src_bufs).  They check for the data pointer being null or not rather than any specific knowledge stemming from the strings_column_view (chars_size()).

**Comment by nvdbaranec** [comment]
Also, I'll probably be retargeting this to 0.18, depending on what Mark thinks.

**Comment by sperlingxx** [comment]
This change LGTM.

**Review by sperlingxx — APPROVED** [review]

**Comment by harrism** [comment]
I think move it to 0.18 due to lack of time to review.

**Review by karthikeyann — APPROVED** [review]

File: cpp/src/copying/contiguous_split.cu
Comment (sperlingxx) [inline cpp/src/copying/contiguous_split.cu:998]: nit: No new line at the end of file

**Review by sperlingxx — APPROVED** [review]
LGTM

**Review by karthikeyann — APPROVED** [review]

**Review by devavret — APPROVED** [review]
approving based on @karthikeyann 's review and to unblock.

**Pull request merged**
